### PR TITLE
[Proposal] Check `NumberParser` with property based testing

### PR DESCRIPTION
**Pull request by dubzzz**
As suggested in issue #2065, using property based testing to check if `NumberParser` does what it says can help detecting issues earlier. I really believe it could be useful to know what is or is not handled by `NumberParser`.

The current PR is a first proposal for the implementation of such test. @snowystinger looked interested by the idea so I wanted to show how it looks like.

Some parts of the tests have been muted as the current implementation of `NumberParser` does not handle them properly:
- percent,
- locale=fr-FR,
- locale=de-DE,
- close to zero,
- far from zero.

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [x] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

In a nutshell: in property based testing you check that a property holds whatever the inputs you pass to your function. Here the property is:

> For any number (d), and options (opts),
> `format(parse(format(d)))` is `format(d)`

## 🧢 Your Project:

https://github.com/dubzzz/fast-check/

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -113,6 +113,7 @@
     "eslint-plugin-react": "^7.21.2",
     "eslint-plugin-react-hooks": "^4.1.2",
     "eslint-plugin-rulesdir": "^0.1.0",
+    "fast-check": "^2.17.0",
```
Comment (dubzzz) [inline package.json:116]: By the way the library is already used by Jest, Jasmine, ramda..

**Comment by snowystinger** [comment]
Awesome, thanks, we'll take a look soon. Thanks for the intro to fc, didn't know what i was missing out on

**Comment by dubzzz** [comment]
Here are just some other notes regarding property based testing.

The **core idea** is that generally speaking, it is not possible to think about all edge cases of an algorithm and to write them all in our test suites (-0, NaN, Infinity, \_\_proto\_\_, MAX_VALUE, MIN_VALUE... or even `{optA: true, opt: false}`, `{optA: false, optB: true}`...). So instead of writing them we ask a framework (here fast-check but they are others depending on your stack) to generate an integer, a string... and it will generate some and assess your property is valid whatever the inputs it tried.

As a consequence it may (and will) not run the exact same tests if you relaunch it. Actually it produces 100 inputs for each call to assert. So is it random? Yes and no. Actually whenever it fails it gives you the seed that caused the run to fail and the path leading to the minimal failing case. Additionally, except in very rare cases where the failure is extremely rare and requires very precise combination of values, it will find the issue in a hundred runs. **Random but deterministic (can be replayed)**

Minimal failing case? Without that it would be difficult to read the output, when asking for a string fast-check tries to build a string but it does not care how readable it will be. Then it runs your property. If it fails it will take the failing string and remove all the unneeded part from it. It is called the shrink. **Make the error human readable and smaller**

One of the key feature is that contrary to homemade random it will tend to generate edge cases more often. So with the generator of double (by default), NaN, Infinity, -0 and MAX_VALUE (which are really rare if we consider all the possible doubles) will fall from time to time. **Edge cases more frequently**

**Comment by snowystinger** [comment]
Hey, I hope you don't mind, I've pushed to your branch.
I began fixing some of the issues that this points out, and I *think* it's getting better
I do have some questions for you about usage of FC, I've commented those in the test file
It does show there are some very odd edge cases out there, so that's great, I'll have some more of the team look at these and see if we or you can spot any generalizations we can do to cut down on coding for the edge cases individually

File: packages/@internationalized/number/test/NumberParser.test.js [hunk 2]
```diff
@@ -10,8 +10,13 @@
  * governing permissions and limitations under the License.
  */
 
+import fc from 'fast-check';
+import messages from '../../../@react-aria/numberfield/intl/*';
 import {NumberParser} from '../src/NumberParser';
 
+// for some reason hu-HU isn't supported in jsdom/node
+let locales = Object.keys(messages).map(locale => locale.replace('.json', '')).filter(locale => locale !== 'hu-HU');
```
Comment (dubzzz) [inline packages/@internationalized/number/test/NumberParser.test.js:18]: Nice, I was looking for a way to get them all! Clearly better than my hard-coded way

File: packages/@internationalized/number/test/NumberParser.test.js [hunk 3]
```diff
@@ -137,10 +142,82 @@ describe('NumberParser', function () {
...
+              fc.pre(opts.minimumSignificantDigits === undefined || opts.maximumSignificantDigits === undefined || opts.minimumSignificantDigits <= opts.maximumSignificantDigits);
+              // invalid currency maximumSignificantDigits, getting an error about too many pre-conditions
+              // fc.pre(!(opts.style === 'currency' && opts.maximumSignificantDigits <= 2));
+              // this set of options for da-DK results in nothing but the unit being displayed, it seems to be a bug in the browser? (new Intl.NumberFormat('da-DK', {"style":"unit","unit":"kilometer-per-hour", 'unitDisplay': 'long', "maximumSignificantDigits":5})).format(0.2)
+              // fc.pre(!(opts.style === 'unit' && opts.unit === 'kilometer-per-hour' && opts.unitDisplay === 'long' && (Math.abs(d) < 1 && Math.abs(d) > 0) && locales === 'da-DK'));
+              const formatter = new Intl.NumberFormat(locales, opts);
+              const parser = new NumberParser(locales, opts);
+
+              // is there some way to combine this with the above logic?
```
Comment (dubzzz) [inline packages/@internationalized/number/test/NumberParser.test.js:207]: Yes you can!

Basically you could do something like (if I understand well the need):
```js
const inputsArb = fc.tuple(valueArb, localesArb, styleOptsArb, genericOptsArb)
  .map(([d, locales, styleOpts, genericOpts]) => {
    const opts = {...styleOpts, ...genericOpts};
    if (styleOpts.style === 'percent') {
      opts.minimumFractionDigits = genericOpts.minimumFractionDigits > 18 ? 18 : genericOpts.minimumFractionDigits;
      opts.maximumFractionDigits = genericOpts.maximumFractionDigits > 18 ? 18 : genericOpts.maximumFractionDigits;
     }
     return {d,opts,locales};
  })
  .filter(({opts}) => opts.minimumFractionDigits === undefined || opts.maximumFractionDigits === undefined || opts.minimumFractionDigits <= opts.maximumFractionDigits)
  .filter(({opts}) => opts.minimumSignificantDigits === undefined || opts.maximumSignificantDigits === undefined || opts.minimumSignificantDigits <= opts.maximumSignificantDigits)
  .map(({d,opts,locales}) => {
    let adjustedNumberForFractions = d;
    if (Math.abs(d) < 1 && opts.minimumFractionDigits && opts.minimumFractionDigits > 1) {
      adjustedNumberForFractions = d * (10 ** (opts.minimumFractionDigits || 2));
    } else if (Math.abs(d) > 1 && opts.minimumFractionDigits && opts.minimumFractionDigits > 1) {
      adjustedNumberForFractions = d / (10 ** (opts.minimumFractionDigits || 2));
    }
    return {adjustedNumberForFractions,opts,locales}
  });

fc.assert(fc.property(
  inputsArb,
  function ({adjustedNumberForFractions, locales, opts}) {
    // no more need for fc.pre or computation of adjustedNumberForFractions
```

File: packages/@internationalized/number/src/NumberParser.ts [hunk 4]
```diff
@@ -191,11 +228,23 @@ class NumberParserImpl {
...
   let singularParts = formatter.formatToParts(1);
+  // some locales such as lv-LV unit inches have words that are specific to 0
+  let zeroParts = formatter.formatToParts(0);
+  // covers sr-SP which has a different word for this (but not 0.1, i clearly don't know how to i18n...) "unit":"liter", "unitDisplay":"long"
+  let decimalLiteralsParts = formatter.formatToParts(0.2);
+  // covers sl-SI which has a different word this (i'm guessing it's the thousands word?) "unit":"inch", "unitDisplay":"long"
+  let decimalLiteralsParts2 = formatter.formatToParts(1000);
+  // covers ru-RU {"style":"unit","unit":"inch"},{"minimumSignificantDigits":2} I DON'T UNDERSTAND THIS ONE!?
+  let foo = formatter.formatToParts(1575645006488261);
+  let bar = formatter.formatToParts(0.971912234434432);
```
Comment (snowystinger) [inline packages/@internationalized/number/src/NumberParser.ts:244]: we need to find a more generalized way of handling these

**Comment by dubzzz** [comment]
@snowystinger Any updates on this one?

**Comment by snowystinger** [comment]
Hey, sorry to say but this fell off my priorities. Last I remember there were a couple of things we still needed to figure out mostly around generalizing. I think where I was struggling with it, was that each case seemed very specific and I was only finding them randomly through these awesome tests. I would love to pick it up again. I think the next step get this going again would be to find the definitions of grammar for each of the locales we support in CLDR for NumberFormatting and determine all the special cases that way instead of through discovery.

**Comment by dubzzz** [comment]
No worry, I was just pinging to get a status. But definitely no urgency given how spicy the issue is

**Comment by snowystinger** [comment]
Closing in favor of https://github.com/adobe/react-spectrum/pull/5156
I didn't want to deal with merges and pushing to a fork ;) the work lives on though.
